src/9FC02700.c

```diff
@@ -1,27 +1,423 @@
+            return -1;
+    }
+}
+
+s32 check_crl_root(BbCrlHead* crlHead, s32 expectedType) {
+    if (crlHead->type != expectedType) {
+        return -1;
+    }
+
+    switch (crlHead->type) {
+        case 0:
+            if (strncmp(crlHead->issuer, aRootXsca, 9) != 0) {
+                return -1;
+            }
+            break;
+        case 1:
+            if (strncmp(crlHead->issuer, aRootCpca, 9) != 0) {
+                return -1;
+            }
+            break;
+        case 2:
+            if (strncmp(crlHead->issuer, aRoot_0, 5) != 0) {
+                return -1;
+            }
+            break;
+        default:
+            return -1;
+    }
+
+    return 0;
+}
+
+s32 check_crlbundle_version(BbCrlBundle* crlBundle) {
+    BbCrlHead* crlHead = crlBundle->head;
+
+    switch (crlHead->type) {
+        case 0:
+            if (D_9FC0F308.tsCrlVersion < crlHead->versionNumber) {
+                D_9FC0F308.tsCrlVersion = crlHead->versionNumber;
+            } else {
+                return 0;
+            }
+            break;
+        case 1:
+            if (D_9FC0F308.cpCrlVersion < crlHead->versionNumber) {
+                D_9FC0F308.cpCrlVersion = crlHead->versionNumber;
+            } else {
+                return 0;
+            }
+            break;
+        case 2:
+            if (D_9FC0F308.caCrlVersion < crlHead->versionNumber) {
+                D_9FC0F308.caCrlVersion = crlHead->versionNumber;
+                D_9FC0F308.tsCrlVersion = D_9FC0F308.contentRlVersion = D_9FC0F308.cpCrlVersion = 0;
+            } else {
+                return 0;
+            }
+            break;
+        default:
+            return -1;
+    }
+
+    if (write_virage01_data(&D_9FC0F308) != 0) {
+        return -1;
+    }
+    
+    return 0;
+}
+
+s32 verify_crlbundle(BbCrlBundle *bundle, u32 type, s32 requiredVersion) {
+    rsaDataBlock dataBlocks[2];
+    u32 expectedVersion;
+    s32 t;
+    s32 ret;
+    BbRsaCert* cert;
+
+    if (get_expected_revocation_list_version(type, &expectedVersion) != 0) {
+        return -1;
+    }
+
+    if (bundle->head == 0) {
+        if (expectedVersion != 0) {
+            return -1;
+        }
+
+        if (requiredVersion != 0) {
+            return -1;
+        }
+
+        return 0;
+    }
+
+    if (expectedVersion > bundle->head->versionNumber) {
+        return -1;
+    }
+
+    if (requiredVersion > bundle->head->versionNumber) {
+        return -1;
+    }
+
+    if (check_crl_root(bundle->head, type) != 0) {
+        return -1;
+    }
+
+    if (type != 2) {
+        switch (type) {
+            case 0:
+                t = 1;
+                break;
+    
+            case 1:
+                t = 2;
+                break;
+    
+            default:
+                return -1;
+        }
+
+        if (verify_cert_chain(bundle->certChain, t)) {
+            return -1;
+        }
+    }
+    
+    dataBlocks[0].data = &bundle->head->type;
+    dataBlocks[0].size = 0x58;
```

Comment:
Is this size expressible in terms of a `sizeof` of something?

Reply:
Yeah I have similar code in func_9FC00BAC

```c
    blocks[0].data = bundle->ticket;
    blocks[0].size = sizeof(BbTicket) - sizeof(BbRsaSig2048);
...
        blocks[0].data = &bundle->ticket->cmd.contentDesc;
        blocks[0].size = sizeof(BbContentDesc);
        blocks[1].data = &contentMetaDataHead;
        blocks[1].size = sizeof(BbContentMetaDataHead) - sizeof(BbRsaSig2048);
```